File: 2021/day25/framework/grid.py
from typing import Iterable, TypeVar, Callable, Optional, Tuple

T = TypeVar('T')
S = TypeVar('S')
class Grid:
    def __init__(self,
                 lines: Iterable[Iterable[S]],
                 conv: Optional[Callable[[S], T]] = None):
        # that is if T and S are the same
        if conv is None:
            conv = lambda x: x
        self._grid = [ [ conv(item) for item in line ] for line in lines]
        self._h = len(self._grid)
        self._w = 0
        self._max_x = None
        self._max_y = None
        if self._h > 0:
            self._w = len(self._grid[0])
            self._max_x = self._h - 1
        if self._w > 0:
            self._max_y = self._w - 1


    @property
    def h(self) -> int:
        return self._h

    @property
    def w(self) -> int:
        return self._w

    @property
    def max_x(self) -> Optional[int]:
        return self._max_x

    @property
    def max_y(self) -> Optional[int]:
        return self._max_y

    def get(self, x: int, y: int) -> T:
        return self._grid[x][y]

    def neighbours(self, x: int, y: int) -> Iterable[Tuple[int, int]]:
        if self._h == 0 or self._w == 0:
            return []
        if x < self._max_x:
            yield (x+1, y)
        if x > 0:
            yield (x-1, y)
        if y < self._max_y:
            yield (x, y+1)
        if y > 0:
            yield (x, y-1)

    def down(self, x: int, y: int, wrap: bool=False)-> Tuple[int,int]:
        if x >= self._max_x:
            if wrap:
                x = 0
            else:
                raise ValueError(f'cannot go down from {x}')
        else:
            x = x+1
        return (x, y)

    def up(self, x: int, y: int, wrap: bool=False)-> Tuple[int,int]:
        if x == 0:
            if wrap:
                x = self._max_x
            else:
                raise ValueError(f'cannot go up from {x}')
        else:
            x = x-1
        return (x, y)

    def right(self, x: int, y: int, wrap: bool=False)-> Tuple[int,int]:
        if y >= self._max_y:
            if wrap:
                y = 0
            else:
                raise ValueError(f'cannot go right from {y}')
        else:
            y = y + 1
        return (x, y)

    def left(self, x: int, y: int, wrap: bool=False)-> Tuple[int,int]:
        if y == 0:
            if wrap:
                y = self._max_y
            else:
                raise ValueError(f'cannot go left from {y}')
        else:
            y = y - 1
        return (x, y)

    def rows(self):
        for row in self._grid:
            yield row

    def row(self, i):
        return self._grid[i]

    def columns(self):
        for j in range(self._max_y):
            yield [ row[j] for row in self._grid ]

    def __iter__(self):
        return self.rows()

    def __str__(self):
        return '\n'.join(
                    ''.join(str(item) for item in row)
                    for row in self.rows()
                )

    def __repr__(self):
        return '\n'.join(
                    ''.join(repr(item) for item in row)
                    for row in self.rows()
                )

class InfiniteGrid:
    def __init__(self,
                 lines: Iterable[Iterable[S]],
                 default: T,
                 conv: Optional[Callable[[S], T]] = None):
        # that is if T and S are the same
        if conv is None:
            conv = lambda x: x
        self._grid = [
            [ conv(item) for item in line ]
            for line in lines
        ]
        self._h = len(self._grid)
        self._w = 0
        self._max_x = None
        self._max_y = None
        if self._h > 0:
            self._w = len(self._grid[0])
            self._max_x = self._h - 1
        if self._w > 0:
            self._max_y = self._w - 1
        self._min_x = 0
        self._min_y = 0
        self._default = default


    @property
    def h(self) -> int:
        return self._h

    @property
    def w(self) -> int:
        return self._w

    @property
    def max_x(self) -> Optional[int]:
        return self._max_x

    @property
    def max_y(self) -> Optional[int]:
        return self._max_y

    def get(self, x: int, y: int) -> T:
        if x > self._max_x or x < 0 or y > self._max_y or y < 0:
            return self._default
        return self._grid[x][y]

    def square(self, x: int, y: int) -> Iterable[Tuple[int, int]]:
        for i in range(x-1, x+2):
            for j in range(y-1, y+2):
                yield self.get(i, j)

    def count(self, p: Callable[[T], bool]) -> int:
        return sum(
            1
            for row in self._grid
            for elm in row
            if p(elm)
        )


    def __str__(self):
        res = []
        for i in range(-1, self._h + 1):
            res.append(
                ''.join(str(self.get(i,j)) for j in range(-1, self._w+1))
            )
        return '\n'.join(res)

File: 2022/framework/grid/grid.go
package grid

import (
	"fmt"
	"strconv"
	"strings"
)

type Point struct {
	x int
	y int
}

type Grid[T any] struct {
	grid [][]T
	h    int
	w    int
	maxX int
	maxY int
}

func NewGrid[T any, S any](lines [][]S, conv func(S) T) Grid[T] {
	var grid [][]T
	h := 0
	w := 0
	maxX := 0
	maxY := 0
	for x, line := range lines {
		if x > maxX {
			maxX = x
			h = x + 1
		}
		grid = append(grid, []T{})
		for y, item := range line {
			if y > maxY {
				maxY = y
				w = y + 1
			}
			grid[x] = append(grid[x], conv(item))
		}
	}

	return Grid[T]{
		grid: grid,
		h:    h,
		w:    w,
		maxX: maxX,
		maxY: maxY,
	}
}

func (g *Grid[T]) H() int {
	return g.h
}

func (g *Grid[T]) W() int {
	return g.w
}

func (g *Grid[T]) MaxX() int {
	return g.maxX
}

func (g *Grid[T]) MaxY() int {
	return g.maxY
}

func (g *Grid[T]) Neighbours(x int, y int) []Point {
	res := []Point{}
	if g.h == 0 || g.w == 0 {
		return res
	}

	if x < g.maxX {
		res = append(res, Point{x + 1, y})
	}

	if x > 0 {
		res = append(res, Point{x - 1, y})
	}

	if y < g.maxY {
		res = append(res, Point{x, y + 1})
	}

	if y > 0 {
		res = append(res, Point{x, y - 1})
	}
	return res
}

func (g *Grid[T]) Down(x, y int, wrap bool) Point {
	if x >= g.maxX {
		if wrap {
			return Point{0, y}
		}
		panic("Cannot go down")
	}

	return Point{x + 1, y}
}

func (g *Grid[T]) Up(x, y int, wrap bool) Point {
	if x == 0 {
		if wrap {
			return Point{g.maxX, y}
		}
		panic("Cannot go up")
	}

	return Point{x - 1, y}
}

func (g *Grid[T]) Right(x, y int, wrap bool) Point {
	if y >= g.maxY {
		if wrap {
			return Point{x, 0}
		}
		panic("Cannot go right")
	}

	return Point{x, y + 1}
}

func (g *Grid[T]) Left(x, y int, wrap bool) Point {
	if y == 0 {
		if wrap {
			return Point{x, g.maxY}
		}
		panic("Cannot go left")
	}

	return Point{x, y - 1}
}

func (g Grid[T]) Rows() [][]T {
	return g.grid
}

func (g Grid[T]) Row(x int) []T {
	return g.grid[x]
}

func (g Grid[T]) Column(y int) []T {
	res := []T{}
	for x := 0; x <= g.maxX; x++ {
		res = append(res, g.grid[x][y])
	}
	return res
}

func (g Grid[T]) Columns() [][]T {
	res := [][]T{}
	for y := 0; y <= g.maxY; y++ {
		res = append(res, g.Column(y))
	}
	return res
}

func Convert[T any, S any](g Grid[T], conv func(T) S) Grid[S] {
	return NewGrid[S, T](g.Rows(), conv)
}

func (g Grid[T]) String() string {
	var lines []string
	gridStr := Convert[T, string](g, func(elm T) string { return fmt.Sprintf("%v", elm) })
	for _, row := range gridStr.Rows() {
		lines = append(lines, strings.Join(row, ""))
	}
	return strings.Join(lines, "\n")
}

func BasicTest() int {
	lines := [][]string{
		[]string{"1", "2", "3"},
		[]string{"4", "5", "6"},
		[]string{"7", "8", "9"},
	}
	g := NewGrid[int, string](lines, func(s string) int {
		e, _ := strconv.Atoi(s)
		return e
	})
	t := NewGrid[int, int](g.Columns(), func(e int) int { return e })
	fmt.Println(g)
	fmt.Println("-------")
	fmt.Println(t)
	return 0
}
